Script template - Add Python annotations

Gustry opened the pull request:
I think it would be nice to have some Python annotations.
It gives an hint for new comers about what are these objects.

If accepted, it would be nice to have in the 3.40.


Side question, what do you think of dropping `self.tr()` method from this template ?
A Processing script cannot be translated I think, the probability that someone will need a translated version of the script is low as well (he would need to go with a normal QGIS plugin).
It would decrease a little bit the comprehension of the "tutorial script" without the `self.tr`.

Gustry committed:
Review the Processing script template : add annotations, remove self.tr

nyalldawson commented:
> Side question, what do you think of dropping self.tr() method from this template ?

Big +1

nyalldawson commented:
While you're updating this, how about changing x createInstance to 

    return self.__class__()

That'll mean we can rename the class in one place only and things will just work

Gustry commented:
> While you're updating this, how about changing createInstance to

Done, and I moved to the end of the file. We do not need to explain it during a workshop.

I also added the example of the icon.

nyalldawson commented on python/plugins/processing/script/ScriptTemplate.py:
```diff
@@ -85,15 +80,25 @@ def groupId(self):
+        # As this example is doing a copy of the input to the output,
+        # we use the "duplicate" layer icon.
+        # You can use "PyQGIS Resource Browser" plugin to find icons
+        return QIcon(":/images/themes/default/mActionDuplicateLayer.svg")
```
I think this format is a bit confusing, and in general using resources for plugins is discouraged in modern PyQt development. (It also is much trickier for PyQt6).

It would be better to use a hardcoded file path instead, as that is immediately understandable for readers, but I realise that it's tricky here! We could get away with a hardcoded invalid path, since that will return an invalid QIcon but not raise any exceptions...

Gustry replied:
What about `QgsApplication.getThemeIcon('mActionDuplicateLayer.svg')` ? Is-it also discouraged ?

For a Python script, it looks indeed tricky to use a hard-coded file path.

What about adding a snippet using a base64 string ? (we can comment this example for instance) :

```
    def icon(self) -> QIcon:
        data = "Your base64 encoded string"
        pixmap = QPixmap()
        pixmap.loadFromData(base64.b64decode(data))
        return QIcon(pixmap)
```

nyalldawson replied:
> What about QgsApplication.getThemeIcon('mActionDuplicateLayer.svg') ? Is-it also discouraged ?

Not discouraged, but also not something I'd think is particularly useful for a user-created algorithm (since you're restricted to the choice of inbuilt icons)

>     def icon(self) -> QIcon:
        data = "Your base64 encoded string"
        pixmap = QPixmap()
        pixmap.loadFromData(base64.b64decode(data))
        return QIcon(pixmap)

If you've got a dummy string there anyway, it's much simpler to just put a dummy hardcoded path there and let the developer work out the logic needed to distribute this icon.

Maybe it's better to just omit icon() in order to keep this template as simple as possible..

Gustry replied:
Ok, I have removed it ;-)

nyalldawson commented on python/plugins/processing/script/ScriptTemplate.py:
```diff
@@ -113,10 +107,15 @@ def initAlgorithm(self, config=None):
-    def processAlgorithm(self, parameters, context, feedback):
+    def processAlgorithm(
+        self,
+        parameters: dict[str, str],
```
```suggestion
        parameters: dict[str, Any],
```

nyalldawson commented on python/plugins/processing/script/ScriptTemplate.py:
```diff
@@ -85,15 +79,15 @@ def groupId(self):
+        return "Example algorithm short description"
 
-    def initAlgorithm(self, config=None):
+    def initAlgorithm(self, config: Optional[dict] = None):
```
```suggestion
    def initAlgorithm(self, config: Optional[dict[str, Any]] = None):
```
